src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
  * actually transmitted matched a pre-computed size and hash
  */
 class Runner {
-  constructor(agent, agentOpts) {
-    this.agent = agent || new https.Agent(agentOpts || {}); 
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      agent: Joi.object(),
+      agentOpts: Joi.object(),
+      maxRetries: Joi.number().min(0).max(10).default(5),
+    }).without('agent', 'agentOpts').optionalKeys(['agent', 'agentOpts']));
+
+    this.agent = opts.agent || new https.Agent(opts.agentOpts || {}); 
+
+    this.maxRetries = opts.maxRetries;
+
   }
 
+  /**
+   * Run a request and retry if a retryable error occurs.  Only 500-series
+   * responses are considered retriable, so any other status code will return
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
```

Comment:
We cannot not support streams for downloading.  If you notice, we don't make any streaming requests of any sort in the `Controller`.  Support for streaming exists so that the `Client` class can do streaming downloads, which is pretty critical.  I think the idea for doing FD based streams is a neat thing to add.  I think supporting that in addition to raw streams and stream factories would be pretty easy to do.

That said, in this exact spot, we don't really need to worry about that, since this is just notifying the user that non-retryable bodies are non-retryable.  We'd need to add that support into the `Runner.prototype.runOnce()` method.